src/app/fdctl/run/tiles/metrics.c

```diff
@@ -0,0 +1,392 @@
+
+  metrics_idx = 0;
+  ulong     num_datapoints = 0;
+  ulong     data_buffer_capacity = data_buffer_sz;
+  Datapoint datapoints[depth];
+  for( ;; ) {
+    long now = fd_tickcount();
+    metrics_t * metric = &metrics[ metrics_idx ];
+
+    /* Periodically post metrics */
+    if( FD_UNLIKELY( metrics_service_deadline <= now ) ) {
+      if( FD_LIKELY( strlen( data_buffer ) ) ) {
+        ssl_send_data( server, port, user_pass, db, post_buffer, post_buffer_sz, data_buffer );
+        data_buffer_capacity = data_buffer_sz;
+        data_buffer[0] = '\0';
+        post_buffer[0] = '\0';
+      }
+      metrics_service_deadline = now + (long)post_interval_ms*1000000 + (long)fd_tempo_async_reload( rng, async_min );
+    }
+
+    /* Potentially read available datapoints from the tile then queue up the next one */
+    for( ulong i = 0; i < depth; i++ ) {
+      metrics_status_t status = metrics_pop_unmanaged( metric, &datapoints[ num_datapoints ] );
+      switch( status ) {
+        case METRICS_STATUS_OK:
+          num_datapoints++;
+          break;
+        case METRICS_STATUS_EMPTY:
+          goto loop_break;
+        case METRICS_STATUS_OVERRUN:
+          FD_LOG_WARNING(( "metrics overrun %s", metrics_tile_names[ metric->tile ] ));
+          break;
+        case METRICS_STATUS_UNINITIALIZED:
+          FD_LOG_WARNING(( "metrics uninitialized %s", metrics_tile_names[ metric->tile ] ));
+          goto loop_break;
+      }
+    }
+    loop_break:
+
+    /* If we have new datapoints collect them now */
+    if( FD_LIKELY( num_datapoints ) ) {
+      data_buffer_capacity = append_datapoints( data_buffer,
+                                                data_buffer_capacity,
+                                                metrics_tile_names[ metric->tile ],
+                                                metric->idx,
+                                                datapoints,
+                                                num_datapoints );
+      num_datapoints = 0;
+
+      /* The buffer is filled, post metrics on the next iteration */
+      if( FD_UNLIKELY( !data_buffer_capacity ) ) {
+        metrics_service_deadline = now;
+      }
+    }
+
+    /* Advance to the next tile */
+    metrics_idx++;
+    metrics_idx %= in_cnt;
+
+    /* Check for CNC signals */
+    if( FD_UNLIKELY( cnc_service_deadline <= now )) {
+      fd_cnc_heartbeat( cnc, now );
+      ulong s = fd_cnc_signal_query( cnc );
+      if( FD_UNLIKELY( s != FD_CNC_SIGNAL_RUN )) {
+        if( FD_LIKELY( s == FD_CNC_SIGNAL_HALT )) break;
+        if( FD_UNLIKELY( s != FD_MUX_CNC_SIGNAL_ACK )) {
+          char buf[FD_CNC_SIGNAL_CSTR_BUF_MAX];
+          FD_LOG_WARNING(( "Unexpected signal %s (%lu) received; trying to resume", fd_cnc_signal_cstr( s, buf ), s ));
+        }
+        fd_cnc_signal( cnc, FD_CNC_SIGNAL_RUN);
+      }
+      cnc_service_deadline = now + ( long ) fd_tempo_async_reload( rng, async_min );
+    }
+  }
+}
+
+static long allow_syscalls[] = {
```

Comment:
I know it's a pain but we can we document why each of these are needed, same as the other tiles? It's very useful for auditing and code review.

Reply:
Updated